server/database.go

```diff
@@ -959,7 +1081,59 @@ func getDatabaseDriver(ctx context.Context, instance *api.Instance, databaseName
 		},
 	)
 	if err != nil {
-		return nil, common.Errorf(common.DbConnectionFailure, fmt.Errorf("failed to connect database at %s:%s with user %q: %w", instance.Host, instance.Port, instance.Username, err))
+		return nil, err
+	}
+
+	return driver, nil
+}
+
+// Try to get database driver using the instance's read-only data source.
+// If no read-only data source is found, try using the admin data source.
+// Upon successful return, caller MUST call driver.Close, otherwise, it will leak the database connection.
+func tryGetReadOnlyDatabaseDriver(ctx context.Context, instance *api.Instance, databaseName string, logger *zap.Logger) (db.Driver, error) {
+	readOnlyDataSource := api.DataSourceFromInstanceWithType(instance, api.RO)
+	if readOnlyDataSource == nil {
+		adminDataSource := api.DataSourceFromInstanceWithType(instance, api.Admin)
+		if adminDataSource == nil {
+			return nil, common.Errorf(common.Internal, fmt.Errorf("not found read-only and admin data source in instance %d", instance.ID))
```

Comment:
```
dataSource := api.DataSourceFromInstanceWithType(instance, api.RO)
if dataSource == nil {
  dataSource = api.DataSourceFromInstanceWithType(instance, api.Admin)
}
if dataSource == nil {
  return nil, common.Errorf(common.Internal, fmt.Errorf("data source not found for instance %d", instance.ID))
}
```

Since we only have read-only and admin data source.